### PR TITLE
3D Grouped Convolutions

### PR DESCRIPTION
**Pull request by affanv14**
Related to https://github.com/Theano/Theano/issues/5621
Tasks:

- [x] Implement AbstractConv version
- [x] Implement CorrMM version
- [x] Implement GPUCorrMM version

### CODE REVIEW

**Comment by nouiz** [comment]
Note cudnn v7 added support for group convolution for 2d and 3d convolution. I think it would be better to don't spend time making conv3d support group convolution for cudnn in this PR. After this PR, you or someone else could update the conv to compile with  cudnn v7 and use their group convolution.

**Comment by nouiz** [comment]
I did the review up to now and it look good.

**Comment by nouiz** [comment]
I checked cudnn v7 group conv doc. It is implemented for 3d and 2d conv for forward and both grad. So the person that will add support for should not loop for num_group and should pass to it the right parameter.

**Comment by affanv14** [comment]
I just checked out the cudnn docs, I should be able to handle that. I can start working on it in a new PR.

**Comment by nouiz** [comment]
Someone else should make tomorrow a PR to make Theano compile with cudnn v7. I would prefer that you wait for this PR to be made. Otherwise it would duplicate the work.

If you have nothing to do, could you do this documentation issue in the mean time? https://github.com/Theano/Theano/issues/5047

Do you think you have finished this PR or there is more change you need to do?

**Comment by affanv14** [comment]
> Someone else should make tomorrow a PR to make Theano compile with cudnn v7. I would prefer that you wait for this PR to be made. Otherwise it would duplicate the work.

Ok, no problem.
> If you have nothing to do, could you do this documentation issue in the mean time?

I still have separable conv3d to do, which shouldn't take more than two days, I can tackle the documentation after that. Then I can pick up cudnn v7 if it is has been merged by then.

> Do you think you have finished this PR or there is more change you need to do?

A few more changes, like skipping old cpu convolution and also skipping cudnn if num_groups is greater than 1

**Comment by affanv14** [comment]
I finished making changes and also did the rebase

**Comment by nouiz** [comment]
jenkins test that

**Comment by nouiz** [comment]
jenkins test this

File: theano/gpuarray/blas.py [hunk 1]
```diff
@@ -1102,11 +1105,12 @@ def pad(self):
...
-        return '%s{%s, %s, %s}' % (
+        return '%s{%s, %s, %s, %s}' % (
             self.__class__.__name__,
             self.border_mode,
             str(self.subsample),
-            str(self.filter_dilation))
+            str(self.filter_dilation),
+            str(self.num_groups))
 
     def flops(self, inp, outp):
```
Comment (nouiz) [inline theano/gpuarray/blas.py]: I forgot, do this method need an update?

Comment (affanv14) [inline theano/gpuarray/blas.py]: I seem to have missed that. I will update that soon for both 2d and 3d.

**Comment by nouiz** [comment]
jenkins test this

**Pull request merged**